### PR TITLE
incusd/network/ovn: Add support to ipv4.dhcp.ranges

### PR DESCRIPTION
**Pull request by jonatas-lima**
## What this PR do?

- Accept a new parameter `ipv4.dhcp.ranges` that will add new entries to OVN `exclude_ips`

### CODE REVIEW

**Comment by stgraber** [comment]
@jonatas-lima thanks for this!

I was about to take over another PR we had which was attempting to fix this #1201 but had been staled lately.

Before I can look at your change, would you mind editing your commit messages to include the required `Signed-off-by: Firstname Lastname <email>` line (see CONTRIBUTING.md)?

**Comment by jonatas-lima** [comment]
> @jonatas-lima thanks for this!
> 
> I was about to take over another PR we had which was attempting to fix this #1201 but had been staled lately.
> 
> Before I can look at your change, would you mind editing your commit messages to include the required `Signed-off-by: Firstname Lastname <email>` line (see CONTRIBUTING.md)?

He is a friend of mine, so a talked to him to continue his work. Thanks for your support! Commits signed.

**Comment by stgraber** [comment]
I believe I fixed the test failures.

File: internal/server/network/driver_ovn.go [hunk 1]
```diff
@@ -1956,14 +1959,54 @@ func (n *ovn) getDHCPv4Reservations() ([]iprange.Range, error) {
 	}
 
 	var dhcpReserveIPv4s []iprange.Range
-	if routerIntPortIPv4 != nil {
-		dhcpReserveIPv4s = []iprange.Range{{Start: routerIntPortIPv4}, {Start: dhcpalloc.GetIP(ipv4Net, -2)}}
+
+	if routerIntPortIPv4 != nil && n.config["ipv4.dhcp.ranges"] != "" {
```
Comment (stgraber) [inline internal/server/network/driver_ovn.go:1963]: That looks wrong.

Prior to this change, any network with IPv4 configured would have its first and last IP addresses reserved. After your change, this reservation now only happens if `ipv4.dhcp.ranges` is non-empty.

File: internal/server/network/driver_ovn.go [hunk 2]
```diff
@@ -1956,14 +1959,54 @@ func (n *ovn) getDHCPv4Reservations() ([]iprange.Range, error) {
...
+		dhcpReserveIPv4s = append(dhcpReserveIPv4s, iprange.Range{Start: lastIP, End: dhcpalloc.GetIP(ipv4Net, -2)})
 	}
 
 	err = UsedByInstanceDevices(n.state, n.Project(), n.Name(), n.Type(), func(inst db.InstanceArgs, nicName string, nicConfig map[string]string) error {
 		ip := net.ParseIP(nicConfig["ipv4.address"])
 		if ip != nil {
-			dhcpReserveIPv4s = append(dhcpReserveIPv4s, iprange.Range{Start: ip})
+			containsIP := false
+			for _, reservedDhcpRange := range dhcpReserveIPv4s {
+				if containsIP = reservedDhcpRange.ContainsIP(ip); containsIP {
```
Comment (stgraber) [inline internal/server/network/driver_ovn.go:2002]: Our normal coding style avoids definitions + test in one statement.
So this should be:

```
containsIP = reservedDhcpRange.ContainsdIP(ip)
if containsIP {
    break
}
```

**Comment by stgraber** [comment]
@jonatas-lima

**Comment by jonatas-lima** [comment]
> @jonatas-lima

Sorry, i was a little busy yesterday. Could you review again?

**Review by stgraber — APPROVED** [review]

**Pull request merged**